azkaban-common/src/main/java/azkaban/flow/FlowUtils.java

```diff
@@ -104,10 +106,27 @@ public static Flow getFlow(final Project project, final String flowName) {
     return flow;
   }
 
-  public static ExecutableFlow createExecutableFlow(final Project project, final Flow flow) {
-    final ExecutableFlow exflow = new ExecutableFlow(project, flow);
-    exflow.addAllProxyUsers(project.getProxyUsers());
-    return exflow;
+  // if pass in executorManagerAdapter, then use it to populate the flow's ExecutionOption
+  public static ExecutableFlow createExecutableFlow(final Project project, final Flow flow,
+   ExecutorManagerAdapter executorManagerAdapter, Logger logger) {
+    ExecutableFlow exFlow = new ExecutableFlow(project, flow);
+    exFlow.addAllProxyUsers(project.getProxyUsers());
+
+    if (executorManagerAdapter == null) {
+      return exFlow;
+    }
+    try {
+      executorManagerAdapter.preloadExecutionOptions(exFlow);
+    } catch (Throwable e) {
+      if(logger != null) {
+        logger.warn("Fail to preload ExecutableFlow, continue without loading ExecutionOptions", e);
+      }
+      // having issue, fallback to without loading ExecutionOptions
+      exFlow = new ExecutableFlow(project, flow);
+      exFlow.addAllProxyUsers(project.getProxyUsers());
```

Comment:
This is duplicate with line 113. I think we can just add the proxy users right before returning the `exFlow` at the end of this method.

Reply:
I think it's safer as we are not sure in the future the previous steps rely on these proxyUsers info

Reply:
Hmmm, i'm lost. Why not returning `return exFlow;` directly but constructing the exFlow again

Reply:
Perhaps the issue would be avoided entirely if we checked if the executorManagerAdapter was null at the top of the stack?

Reply:
This line 125/126 is to avoid the case that `preloadExecutionOptions()` messed up the exFlow then throw the exception